### PR TITLE
Add org.godotengine.GodotSharp (Godot 4 with C#/.NET support)

### PR DESCRIPTION
**Pull request by Zishan-Rahman**
### Please confirm your submission meets all the criteria

- [X] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [X] My pull request follows the instructions at [App Submission][submission].
- [X] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [X] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [X] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:** https://github.com/flathub/org.godotengine.Godot/pull/109#issuecomment-1724624274
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [X] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission

### CODE REVIEW

**Comment by hfiguiere** [comment]
why a new package?

**Comment by Zishan-Rahman** [comment]
> why a new package?

@hfiguiere This Flatpak originates from a PR opened by @CharlieQLe last year (flathub/org.godotengine.godot#109). Currently, Godot ships with or without C# support, and, while the `.tar.xz` source code file used to build the standard Godot 3 and 4 Flatpaks _can_ be used to build the C#/.NET runtime, both have different dependencies and also operate differently in the way Godot projects are created and exported.

Charlie's PR is currently unmaintained, so [I asked him about the possibility of getting Flatpaks for _both_ Godot 3 _and_ 4 with C# support added to Flathub](https://github.com/flathub/org.godotengine.Godot/pull/109#issuecomment-1724624274), and he gave me the go-ahead in his reply:

> @Zishan-Rahman Go for it! A separate C# flatpak would be a good thing to see for people just getting into Godot that need a Flatpak, though it should be noted that the custom dotnet editors isn't supported afaik.

@Calinou himself also agrees that hosting a .NET Godot Flatpak separately from the standard version already up on Flathub would be a good idea, [as he says so in this comment](https://github.com/flathub/org.godotengine.Godot/issues/8#issuecomment-1455196039):

> **Note that this will have to be a separate Flatpak, as C# support not only makes the Flatpak larger but also requires all projects exported with it to contain the .NET runtime (which makes your exported project larger). This is problematic for projects that only use GDScript.**

I opened a PR for adding the Mono version of Godot 3 to Flathub too: #4595. As well as Godot 3 still being widely used in production (both standard and C#/.NET versions), Godot 4's C#/.NET version, in its current state, cannot target mobile and web platforms, whereas Godot 3's C#/.NET version can. Godot 3 still receives updates as well, as version 3.5.3 was released on late September, and version 3.6 is also in development and set to be released later this year.

File: org.godotengine.GodotSharp.yaml [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+      builtin_harfbuzz=no
+      udev=no
+finish-args:
+  - --share=ipc
+  - --socket=x11
+  - --share=network
+  - --socket=pulseaudio
+  - --filesystem=host
+  - --device=all
+  - --talk-name=org.freedesktop.Flatpak
```
Comment (bbhtt) [inline org.godotengine.GodotSharp.yaml:64]: This needs an exception in the linter

File: org.godotengine.GodotSharp.appdata.xml [hunk 2]
```diff
@@ -0,0 +1,76 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+  <id>org.godotengine.GodotSharp.desktop</id>
```
Comment (bbhtt) [inline org.godotengine.GodotSharp.appdata.xml:3]:
```suggestion
  <id>org.godotengine.GodotSharp</id>
```

File: org.godotengine.GodotSharp.yaml [hunk 3]
```diff
@@ -0,0 +1,141 @@
+app-id: org.godotengine.GodotSharp
```
Comment (bbhtt) [inline org.godotengine.GodotSharp.yaml:1]: Shouldn't this be `org.godotengine.Godot4Sharp`?

Comment (hfiguiere) [reply org.godotengine.GodotSharp.yaml:1]: So how does it work with distro installed version? Seems like a bug if it dumps dotNet into the product even if not used.

**Comment by Zishan-Rahman** [comment]
> So how does it work with distro installed version? Seems like a bug if it dumps dotNet into the product even if not used.

The C#/.NET version has its own export templates to deal with projects using C# code that need to bundle the .NET runtime. These templates are not included with Godot usually, Flatpak or otherwise, and are downloaded through the Godot editor when it is time to export your game (same goes with the normal version which downloads its own templates).

Furthermore, in my own testing, I tried exporting an existing GDScript-only Godot 4 project with both the standard build and the C#/.NET build of Godot 4, and the difference in the size of the executable files (`.x86_64` in my case) was minor (70.5MB for the one made through the C#/.NET version vs 70.3MB for the one made through the standard version), and the `.pck` files generated by both exports were the exact same size. The only time new folders are created for the exported project, at least with Godot 4's .NET version, is when the project contains C# code (and, therefore, also contains `.sln` and `.csproj` files). When I created a test project with an empty C# script and exported that, the total size of all the exported files came to 146.7 MB. When creating and working on a project with Godot 4's C#/.NET version, the `.sln` and `.csproj` files (and the need to build the project through .NET) only gets created when a C# script is created. Otherwise, if one decides to only use GDScript, they can do so and export normally.

@Calinou I apologise for the ping, but if you could please chime in with anything else and correct me if you need to, I'd very much appreciate it.

Sorry for the late reply, by the way.

File: org.godotengine.GodotSharp.yaml [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    buildsystem: simple
+    cleanup: ['*']
+    sources:
+      - type: archive
+        sha256: 3d43b2303a924816ea0e1b345ff04c9b3e27b53eadf0f26012fc0c29b019685f
+        url: https://downloads.sourceforge.net/project/scons/scons/4.4.0/SCons-4.4.0.tar.gz
+    build-commands:
+      - pip3 install --no-index --no-build-isolation --prefix=/app .
+
+  - name: godot-tools
```
[the remark this answers is not in the document]
Comment (Zishan-Rahman) [reply org.godotengine.GodotSharp.yaml:95]: I have to update `nuget-sources.json`. I will comment here again when I have done so.

[the remark this answers is not in the document]
Comment (Zishan-Rahman) [reply org.godotengine.GodotSharp.yaml:95]: I just updated `Microsoft.NETCore.App.Ref` and `Microsoft.AspNetCore.App.Ref` in `nuget-sources.json` to 6.0.23: e7e541fe799c7feba06aeefe2ec0cb582b26d983

File: org.godotengine.GodotSharp.appdata.xml
Comment (bbhtt) [inline org.godotengine.GodotSharp.appdata.xml:1]:
```
org.godotengine.GodotSharp.appdata.xml: FAILED:
\u2022 tag-invalid           : <url> type invalid [unknown]
\u2022 style-invalid         : <li> cannot contain a hyperlink [External script editors are still supported (for use outside of Mono), but you need to follow the steps described here: https://github.com/flathub/org.godotengine.GodotSharp#using-an-external-script-editor]
\u2022 style-invalid         : <li> cannot contain a hyperlink [Godot 4's Blender importer is supported, but Blender needs to be installed on your system and you need to follow the steps described here: https://github.com/flathub/org.godotengine.GodotSharp#using-blender]
Validation of files failed
```

File: org.godotengine.GodotSharp.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (bbhtt) [reply org.godotengine.GodotSharp.yaml:95]: It builds now

File: org.godotengine.GodotSharp.appdata.xml
Comment (Zishan-Rahman) [reply org.godotengine.GodotSharp.appdata.xml:1]: The standard version of Godot that's up on Flathub has URLs in its appdata.xml, though, and that goes through without issue (see below). Maybe Buildbot ignores them or lets them through when building Flatpaks?
Screenshot from 2023-10-13 12-10-39

Comment (bbhtt) [reply org.godotengine.GodotSharp.appdata.xml:1]: Flathub uses a patched appstream-util that seems to be cool with it but not the upstream one. 

Should be fine, I keep forgetting to use the patched one.

**Comment by Zishan-Rahman** [comment]
I've amended https://github.com/flathub/flatpak-builder-lint/pull/190 to include an exception for _only_ GodotSharp. I plan to return to Godot3Sharp once _this_ one's up on Flathub. [I'm currently having some weird problems trying to build Godot3Sharp locally on my own computer](https://github.com/flathub/flathub/pull/4595#discussion_r1358159712), and I'm hoping these problems don't persist on a test build done by Buildbot. I'll return to Godot3Sharp once GodotSharp (this one) is up on Flathub.

**Comment by Zishan-Rahman** [comment]
OK, the flatpak-builder-lint PR has just been merged. Now, I think we're ready for a test build. It's been long enough.

bot, build org.godotengine.GodotSharp

**Comment by Zishan-Rahman** [comment]
bot, build org.godotengine.GodotSharp

**Comment by Zishan-Rahman** [comment]
I tried opening [vaporvee's C# third-person controller project template](https://github.com/vaporvee/gd-net-thirdpersoncontroller) in the test Flatpak, and, after building the initial C# solution (using `Project --> Tools --> C# --> Create C# Solution`), the project loads and runs as it should. I am also able to create a Godot project and add a C# script to it, before building the C# solution and then running _that_ project too. This looks like it's ready to _finally_ add to Flathub :+1: 

Now I'll go back to [org.godotengine.Godot3Sharp](https://github.com/flathub/flathub/pull/4595) and try to get _that_ working as well. I'm more than happy to maintain GodotSharp and Godot3Sharp simultaneously. Please can @Calinou and @CharlieQLe be added to GodotSharp as maintainers (and also to Godot3Sharp when _that_ gets added to Flathub)?

**Pull request closed without merging**